src/deploy/functions/params.ts

```diff
@@ -0,0 +1,333 @@
+ *   - if the default CEL expression is not resolvable--it depends on a param whose value is not yet known--we throw an error
+ *   - yes, this means that the same set of params may or may not throw depending on the order the SDK provides them to us in
+ *   - after prompting, the resolved value of the param is written to the most specific .env file available
+ */
+export async function resolveParams(
+  params: Param[],
+  projectId: string,
+  userEnvs: Record<string, ParamValue>
+): Promise<Record<string, ParamValue>> {
+  const paramValues: Record<string, ParamValue> = {};
+
+  for (const param of params) {
+    if (userEnvs.hasOwnProperty(param.param)) {
+      if (canSatisfyParam(param, userEnvs[param.param])) {
+        paramValues[param.param] = userEnvs[param.param];
+        continue;
+      } else {
+        throw new FirebaseError("");
+      }
+    }
+
+    if (param.default) {
+      let paramDefault: ParamValue;
+      if (isCEL(param.default)) {
+        paramDefault = resolveDefaultCEL(param.type || "", param.default, paramValues);
+      } else {
+        paramDefault = param.default;
+      }
+      paramValues[param.param] = await promptParam(param, paramDefault);
+    } else {
+      paramValues[param.param] = await promptParam(param);
+    }
+  }
+
+  return paramValues;
+}
+
+/**
+ * Returns the resolved value of a user-defined Functions parameter.
+ * Functions params are defined by the output of the Functions SDK, but their value is not set until deploy-time.
+ *
+ * For most param types, we check the contents of the dotenv files first for a matching key, then interactively prompt the user.
+ * When the CLI is running in non-interactive mode or with the --force argument, it is an error for a param to be undefined in dotenvs.
+ */
+async function promptParam(param: Param, resolvedDefault?: ParamValue): Promise<ParamValue> {
+  if (resolvedDefault !== undefined && !canSatisfyParam(param, resolvedDefault)) {
+    throw new FirebaseError("");
+  }
+
+  switch (param.type) {
+    case "string":
+      return promptStringParam(param, resolvedDefault as string);
+    case "int":
+      return promptIntParam(param, resolvedDefault as number);
+    default:
+      throw new FirebaseError("Build specified parameter " + param + " with unsupported type");
+  }
+}
+
+async function promptStringParam(param: StringParam, resolvedDefault?: string): Promise<string> {
+  if (!param.input) {
+    const defaultToText: TextInput<string> = { text: {} };
+    param.input = defaultToText;
+  }
+
+  switch (param.input.type) {
+    case "select":
+      throw new FirebaseError(
+        "Build specified string parameter " + param.param + " with unsupported input type 'select'"
+      );
+    case "hardcoded":
+      return resolvedDefault || "";
+    case "text":
+    default:
+      let prompt = `Enter a value for ${param.label || param.param}:`;
```

Comment:
let's always mention the param name and then add the label as optional if it exists.

E.g.

```
let prompt = `Enter a value for ${param.param}`
if (param.label) {
  prompt = `${prompt} (${param.label})`
}
```


(or vice versa; don't know which should be in the parens.